File: foundry/module/scripts/srd/exhaustion.js
/** SRD 5.2 (2024) one-track exhaustion — mirrors `ai_dm.rules.exhaustion`. */
import { srd } from "./core_loader.js";

function cfg() {
    return srd()?.exhaustion ?? {
        max_level: 6, death_at: 6,
        per_level: { d20_penalty: -2, speed_penalty_ft: -5 },
    };
}

export function clampLevel(level) {
    const max = cfg().max_level;
    return Math.max(0, Math.min(max, level | 0));
}

export function add(level, n = 1) { return clampLevel((level | 0) + (n | 0)); }
export function remove(level, n = 1) { return clampLevel((level | 0) - (n | 0)); }

export function d20Penalty(level) {
    return clampLevel(level) * cfg().per_level.d20_penalty;
}
export function speedPenalty(level) {
    return clampLevel(level) * cfg().per_level.speed_penalty_ft;
}
export function isDead(level) { return clampLevel(level) >= cfg().death_at; }

File: src/ai_dm/rules/exhaustion.py
"""SRD 5.2 (2024) one-track exhaustion.

Each level imposes ``-2`` to all d20 tests (checks, attacks, saves) and
``-5 ft`` of speed. Level 6 = death.
"""
from __future__ import annotations

from ai_dm.rules.srd_core import load

_DATA = load("exhaustion")
MAX_LEVEL: int = int(_DATA["max_level"])
DEATH_AT: int = int(_DATA["death_at"])
_PER = _DATA["per_level"]
D20_PENALTY_PER_LEVEL: int = int(_PER["d20_penalty"])
SPEED_PENALTY_PER_LEVEL_FT: int = int(_PER["speed_penalty_ft"])


def clamp(level: int) -> int:
    return max(0, min(MAX_LEVEL, int(level)))


def add(level: int, n: int = 1) -> int:
    return clamp(int(level) + int(n))


def remove(level: int, n: int = 1) -> int:
    return clamp(int(level) - int(n))


def d20_penalty(level: int) -> int:
    """Negative bonus applied to every d20 test."""
    return D20_PENALTY_PER_LEVEL * clamp(level)


def speed_penalty(level: int) -> int:
    """Negative speed adjustment in feet."""
    return SPEED_PENALTY_PER_LEVEL_FT * clamp(level)


def is_dead(level: int) -> bool:
    return clamp(level) >= DEATH_AT


__all__ = [
    "D20_PENALTY_PER_LEVEL",
    "DEATH_AT",
    "MAX_LEVEL",
    "SPEED_PENALTY_PER_LEVEL_FT",
    "add",
    "clamp",
    "d20_penalty",
    "is_dead",
    "remove",
    "speed_penalty",
]

